### DISCUSSION TITLE
It is possible to build Odoo Docker image for darwin/arm64 Macbook with M1 CPU?

### DISCUSSION THREAD

**Discussion by vasi26ro** [Tecnativa/doodba#506] [Ideas]
Hello,
I use doodba, and it is a really impressive tool. Recently I have changed my laptop with a Macbook with M1 CPU. 
I have tried to set up a development environment on this laptop, but I cannot run the odoo image since there is a limitation regading inotify and qemu.

> However, attempts to run Intel-based containers on Apple silicon machines under emulation can crash as qemu sometimes fails to run the container. In addition, filesystem change notification APIs (inotify) do not work under qemu emulation. Even when the containers do run correctly under emulation, they will be slower and use more memory than the native equivalent.

Source: https://docs.docker.com/desktop/mac/apple-silicon/

How can I build the image for M1?

**Comment by piqoni** [comment]
In which image do you get the error?

**Reply by vasi26ro** [reply]
In the odoo image

```
development-env-odoo-1                        |     self._i = Inotify(block_duration_s=block_duration_s)
development-env-odoo-1                        |   File "/usr/local/lib/python3.8/site-packages/inotify/adapters.py", line 57, in __init__
development-env-odoo-1                        |     self.__inotify_fd = inotify.calls.inotify_init()
development-env-odoo-1                        |   File "/usr/local/lib/python3.8/site-packages/inotify/calls.py", line 34, in _check_nonnegative
development-env-odoo-1                        |     raise InotifyError("Call failed (should not be -1): (%d)" % 
development-env-odoo-1                        | inotify.calls.InotifyError: Call failed (should not be -1): (-1) ERRNO=(0)
development-env-odoo-1                        | Exception ignored in: <function Inotify.__del__ at 0x4007c65940>
development-env-odoo-1                        | Traceback (most recent call last):
development-env-odoo-1                        |   File "/usr/local/lib/python3.8/site-packages/inotify/adapters.py", line 79, in __del__
development-env-odoo-1                        |     os.close(self.__inotify_fd)
development-env-odoo-1                        | AttributeError: 'Inotify' object has no attribute '_Inotify__inotify_fd'
```

**Reply by edipiqoni** [reply]
I remember I have circumvented  this one by not using reload/inotify option. (maybe its called "dev mode", not sure now)

Basically it fails in this line in server.py:
![image](https://user-images.githubusercontent.com/123366148/218026935-8993ffc4-3daf-4c4b-82f8-026f1a2218fb.png)

For the root cause, here's an issue (and related issues) which could be a starting point if you want to investigate further: https://github.com/docker/for-mac/issues/6174

**Comment by jigypeper** [comment]
Hi, did you ever get a fix for running the container on apple silicon? I am getting :

[+] Running 2/7
 ✘ odoo_proxy Error                                                                                                                                              5.1s
 ✘ proxy_www_google_com Error                                                                                                                        5.1s
 ⠙ proxy_cdnjs_cloudflare_com Pulling                                                                                                              5.1s
 ⠙ proxy_fonts_googleapis_com Pulling                                                                                                             5.1s
 ⠙ proxy_fonts_gstatic_com Pulling                                                                                                                    5.1s
 ⠙ proxy_www_googleapis_com Pulling                                                                                                              5.1s
 ⠙ proxy_www_gravatar_com Pulling                                                                                                                   5.1s
no matching manifest for linux/arm64/v8 in the manifest list entries

**Comment by vasi26ro** [comment]
> Hi, did you ever get a fix for running the container on apple silicon? I am getting :
> 
> 
> 
> [+] Running 2/7
> 
>  ✘ odoo_proxy Error                                                                                                                                              5.1s
> 
>  ✘ proxy_www_google_com Error                                                                                                                        5.1s
> 
>  ⠙ proxy_cdnjs_cloudflare_com Pulling                                                                                                              5.1s
> 
>  ⠙ proxy_fonts_googleapis_com Pulling                                                                                                             5.1s
> 
>  ⠙ proxy_fonts_gstatic_com Pulling                                                                                                                    5.1s
> 
>  ⠙ proxy_www_googleapis_com Pulling                                                                                                              5.1s
> 
>  ⠙ proxy_www_gravatar_com Pulling                                                                                                                   5.1s
> 
> no matching manifest for linux/arm64/v8 in the manifest list entries

This error is because there is no image build for linux/arm64/v8 platform.

You can build a image for this platform or use the platform: linux/amd64 keyword in docker-compose. The latter option on my laptop was slow.

**Reply by jigypeper** [reply]
Sorry, i'm totally new to docker, and the dev environment at my company is pretty much linux, so it all works fine on there. How would I build this? I tried adding the platform to all the services in the docker compose file, but it brings up this message, maybe i missed one?
'image with reference axllent/mailpit was found but does not match the specified platform: wanted linux/amd64, actual: linux/arm64'

**Reply by jigypeper** [reply]
@vasi26ro any quick tips on how to build this image for Arm64/v8? (sorry, i've got minimal docker experience)
Currently we are using invoke commands to automate the image builds and running the containers.

**Reply by edipiqoni** [reply]
Personally I have just removed those proxy images from my docker-file as they were not needed in my use case, so that would be one way to easily circumvent it.

**Reply by jigypeper** [reply]
as well as rebuilding for arm?

**Reply by theangryangel** [reply]
@jigypeper reach out to me on Monday please and I will try and help explain the problem(s) you're seeing, and why we've not put any effort into this yet.

**Reply by jigypeper** [reply]
I will do!!

**Reply by bosd** [reply]
There is a pr in: https://github.com/Tecnativa/docker-whitelist/pull/15 :)

**Reply by jigypeper** [reply]
Switching to orbstack fixed the issues I was having.
Could carry on development as before.
